### PR TITLE
Jetty 9.4.x 2140 infinispan expired sessions

### PR DESCRIPTION
**Pull request by janbartel**
#2140 

Adds the ability to Infinispan and Hazelcast to query to find expired sessions, as per the other SessionDataStores. This means that sessions can be scavenged by jetty (rather than relying on configuring infinispan or hazelcast with time-to-live on each entry), so HttpSessionListeners can be correctly called when the sessions are scavenged.

At the same time, infinispan has been updated to version 9.4.8, and support for the old infinispan-7 dropped.

### CODE REVIEW

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 1]
```diff
@@ -184,7 +187,52 @@ public boolean isPassivating()
             }
             return false;
         } ).collect( Collectors.toSet() );
+        
+        
+        //Now find other sessions in hazelcast that have expired
+        final AtomicReference<Set<String>> reference = new AtomicReference<>();
+        final AtomicReference<Exception> exception = new AtomicReference<Exception>();
```
Comment (olamy) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: trivial change to ` new AtomicReference<>(); `

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 2]
```diff
@@ -184,7 +187,52 @@ public boolean isPassivating()
...
+        
+        Runnable r = new Runnable()
```
Comment (olamy) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: what about using a lambda here?

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 3]
```diff
@@ -184,7 +187,52 @@ public boolean isPassivating()
...
+        {
+            @Override
+            public void run ()
+            {
+                try
+                {
+                    Set<String> ids = new HashSet<>();
+                    EntryObject eo = new PredicateBuilder().getEntryObject();
+                    Predicate<?, ?> predicate = eo.get("expiry").greaterThan(0).and(eo.get("expiry").lessEqual(now));
```
Comment (olamy) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: Good one. I wonder if we could use that somewhere else in the code (but yes part of an other change :) )

File: tests/test-sessions/test-infinispan-sessions/src/test/java/org/eclipse/jetty/server/session/ClusteredOrphanedSessionTest.java [hunk 4]
```diff
@@ -22,6 +22,7 @@
 import org.eclipse.jetty.session.infinispan.InfinispanSessionDataStoreFactory;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
```
Comment (olamy) [inline tests/test-sessions/test-infinispan-sessions/src/test/java/org/eclipse/jetty/server/session/ClusteredOrphanedSessionTest.java]: unused import

File: tests/test-sessions/test-infinispan-sessions/src/test/java/org/eclipse/jetty/server/session/remote/RemoteClusteredInvalidationSessionTest.java [hunk 5]
```diff
@@ -24,6 +24,7 @@
 import org.eclipse.jetty.session.infinispan.InfinispanSessionDataStoreFactory;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
```
Comment (olamy) [inline tests/test-sessions/test-infinispan-sessions/src/test/java/org/eclipse/jetty/server/session/remote/RemoteClusteredInvalidationSessionTest.java]: unused import

File: tests/test-sessions/test-infinispan-sessions/src/test/java/org/eclipse/jetty/server/session/remote/RemoteClusteredSessionScavengingTest.java [hunk 6]
```diff
@@ -24,6 +24,8 @@
 import org.eclipse.jetty.session.infinispan.InfinispanSessionDataStoreFactory;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
```
Comment (olamy) [inline tests/test-sessions/test-infinispan-sessions/src/test/java/org/eclipse/jetty/server/session/remote/RemoteClusteredSessionScavengingTest.java]: unused import

File: tests/test-sessions/test-infinispan-sessions/src/test/java/org/eclipse/jetty/server/session/remote/RemoteInfinispanTestSupport.java [hunk 7]
```diff
@@ -67,6 +79,22 @@
...
+            
+            ByteArrayOutputStream baos;
+            try(InputStream is = RemoteInfinispanSessionDataStoreTest.class.getClassLoader().getResourceAsStream("session.proto"))
+            {
+                if (is == null)
+                    throw new IllegalStateException("inputstream is null");
+                
+                baos = new ByteArrayOutputStream();
+                IO.copy(is, baos);
+                is.close();
```
Comment (olamy) [inline tests/test-sessions/test-infinispan-sessions/src/test/java/org/eclipse/jetty/server/session/remote/RemoteInfinispanTestSupport.java]: no need to close

**Comment by janbartel** [comment]
I've just found an issue with hazelcast, so don't merge this PR yet until I can resolve it. Basically, the problem is that if I enable indexing on the SessionData.expiry field, then for some reason, hazelcast then uses a separate thread to re-read the SessionData after storing it! So far I haven't been able to find a way to avoid this - I may have to make it configurable whether to support queries on hazelcast, with the documented caveat that if you do, you won't be able to store session attributes of classes that exist only inside the webapp.

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java (hunk 2 above)
Comment (janbartel) [reply jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: Done

**Pull request merged**